text/0000-fragment-specifiers-for-generic-arguments.md

```diff
@@ -0,0 +1,166 @@
+- Feature Name: fragment-specifiers-for-generic-arguments
+- Start Date: 2023-05-31
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Right now, there is no support for parsing the syntax of generic
+parameters/arguments in declarative macros. This makes it difficult to
+impossible to write a declarative macro that handles arbitrary generics easily.
+I propose adding fragment specifiers to parse a generic parameter definition as
+a whole and also parts of the definition.
+
+# Motivation
+[motivation]: #motivation
+
+I personally encountered this issue when attempting to write a declarative macro
+to implement a trait on a type. I wanted to write something like this minimal
+toy example:
+
+```rust
+macro_rules! implement_debug {
+    { $params:generic $type:ty $( where $where:where_clause )? } => {
+        impl $params Debug for $ty $( where $where )? {
+            /* .. implementation goes here .. */
+        }
+    };
+}
+
+struct Container<'a, T>(&'a T);
+implement_debug!(<'a, T: Debug + 'a> Container<'a, T>);
+
+struct Container2<'a, T>(&'a T);
+implement_debug!(<'a, T> Container<'a, T> where T: Debug + 'a);
+```
+
+However, with the current state of declarative macros, there's no way to parse
+arbitrary generic parameters in the body of the macro, forcing this macro of
+mine to be a procedural macro. However, declarative macros are easier to read
+and write, and this could be a declarative macro if there was only a way to
+parse generic parameters.
+
+Additionally, more complicated macros want to be able to parse each parameters
+and its bounds for use in various places, so I'd like this as well:
+
+```rust
+macro_rules! implement_debug {
+    { < $( $param:generic_param $( : $bound:generic_bounds )? ),+ > $type:ty } => {
+        impl < $( $param $( : $bound )? ),+ > Debug for $ty {
+            /* .. implementation goes here .. */
+        }
+    };
+}
+
+struct Container<'a, T>(&'a T);
+
+implement_debug!(<'a, T: Debug + 'a> Container<'a, T>);
+```
+
+Any toy example will be obviously redundant with `:generic`, but more involved
+macros sometimes want it.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When explaining fragment specifiers, this can be explained by adding the new
+fragment specifiers and their descriptions:
+
+* `:generic`: A full set of generic parameters and their bounds (e.g. `<'a, T:
```

Comment:
maybe a nit but `generic` being singular confuses me. I think it would make more sense for it to be named `generics` or `generic_set`, or anything to add plurality to the concept, especially since it would be responsible for parsing the opening and closing angle brackets